F401 false positive (imported but unused)

```python
from functools import reduce

import click

if click.confirm("do you want to disable reduce?", default=False):

    def reduce(f, d):  # noqa: F811
        return d


print(reduce(lambda x, y: x + y, [1, 2, 3, 4, 5]))
```

It is supposed to work like this:

```
snippets » python w0611.py
do you want to disable reduce? [y/N]:
15
snippets » python w0611.py
do you want to disable reduce? [y/N]: y
[1, 2, 3, 4, 5]
```

Here is what ruff outputs:

```
> ruff w0611.py
w0611.py:1:23: F401 `functools.reduce` imported but unused
Found 1 error(s).
1 potentially fixable with the --fix option.
```

If we apply the fix, then the code no longer works if the user presses "N" to not disable reduce.

```
> ruff --version
ruff 0.0.224
```

--- a/resources/test/fixtures/pyflakes/multi_statement_lines.py
+++ b/resources/test/fixtures/pyflakes/multi_statement_lines.py
@@ -1,51 +1,51 @@
 
 if True:
-    import foo; x = 1
-    import foo;     x = 1
+    import foo1; x = 1
+    import foo2;     x = 1
 
 if True:
-    import foo; \
+    import foo3; \
 x = 1
 
 if True:
-    import foo \
+    import foo4 \
         ; x = 1
 
 
 if True:
-    x = 1; import foo
+    x = 1; import foo5
 
 
 if True:
     x = 1; \
-         import foo
+         import foo6
 
 
 if True:
     x = 1 \
-        ; import foo
+        ; import foo7
 
 
 if True:
-    x = 1; import foo; x = 1
-    x = 1;     import foo;     x = 1
+    x = 1; import foo8; x = 1
+    x = 1;     import foo9;     x = 1
 
 if True:
     x = 1; \
-        import foo; \
+        import foo10; \
     x = 1
 
 if True:
     x = 1 \
-        ;import foo \
+        ;import foo11 \
         ;x = 1
 
 
 # Continuation, but not as the last content in the file.
 x = 1; \
-import foo
+import foo12
 
 # Continuation, followed by end-of-file. (Removing `import foo` would cause a syntax
 # error.)
 x = 1; \
-import foo
+import foo13

--- a/resources/test/fixtures/ruff/RUF100_1.py
+++ b/resources/test/fixtures/ruff/RUF100_1.py
@@ -1,64 +1,89 @@
-# This should ignore both errors.
-from typing import (  # noqa: F401
-    List,
-    Sequence,
-)
+def f():
+    # This should ignore both errors.
+    from typing import (  # noqa: F401
+        List,
+        Sequence,
+    )
 
-# This should ignore both errors.
-from typing import (  # noqa
-    List,
-    Sequence,
-)
 
-# This should ignore both errors.
-from typing import (
-    List,  # noqa: F401
-    Sequence,  # noqa: F401
-)
+def f():
+    # This should ignore both errors.
+    from typing import (  # noqa
+        List,
+        Sequence,
+    )
 
-# This should ignore both errors.
-from typing import (
-    List,  # noqa
-    Sequence,  # noqa
-)
 
-# This should ignore the first error.
-from typing import (
-    Mapping,  # noqa: F401
-    Union,
-)
+def f():
+    # This should ignore both errors.
+    from typing import (
+        List,  # noqa: F401
+        Sequence,  # noqa: F401
+    )
 
-# This should ignore both errors.
-from typing import (  # noqa
-    List,
-    Sequence,
-)
 
-# This should ignore the error, but the inner noqa should be marked as unused.
-from typing import (  # noqa: F401
-    Optional,  # noqa: F401
-)
+def f():
+    # This should ignore both errors.
+    from typing import (
+        List,  # noqa
+        Sequence,  # noqa
+    )
 
-# This should ignore the error, but the inner noqa should be marked as unused.
-from typing import (  # noqa
-    Optional,  # noqa: F401
-)
 
-# This should ignore the error, but mark F501 as unused.
-from typing import (  # noqa: F401
-    Dict,  # noqa: F501
-)
+def f():
+    # This should ignore the first error.
+    from typing import (
+        Mapping,  # noqa: F401
+        Union,
+    )
 
-# This should ignore the error, but mark F501 as unused.
-from typing import (  # noqa: F501
-    Tuple,  # noqa: F401
-)
 
-# This should ignore both errors.
-from typing import Any, AnyStr  # noqa: F401
+def f():
+    # This should ignore both errors.
+    from typing import (  # noqa
+        List,
+        Sequence,
+    )
 
-# This should ignore both errors.
-from typing import AsyncIterable, AsyncGenerator  # noqa
 
-# This should mark F501 as unused.
-from typing import Awaitable, AwaitableGenerator  # noqa: F501
+def f():
+    # This should ignore the error, but the inner noqa should be marked as unused.
+    from typing import (  # noqa: F401
+        Optional,  # noqa: F401
+    )
+
+
+def f():
+    # This should ignore the error, but the inner noqa should be marked as unused.
+    from typing import (  # noqa
+        Optional,  # noqa: F401
+    )
+
+
+def f():
+    # This should ignore the error, but mark F501 as unused.
+    from typing import (  # noqa: F401
+        Dict,  # noqa: F501
+    )
+
+
+def f():
+    # This should ignore the error, but mark F501 as unused.
+    from typing import (  # noqa: F501
+        Tuple,  # noqa: F401
+    )
+
+
+def f():
+    # This should ignore both errors.
+    from typing import Any, AnyStr  # noqa: F401
+
+
+def f():
+    # This should ignore both errors.
+    from typing import AsyncIterable, AsyncGenerator  # noqa
+
+
+def f():
+    # This should mark F501 as unused.
+    from typing import Awaitable, AwaitableGenerator  # noqa: F501

--- a/src/checkers/ast.rs
+++ b/src/checkers/ast.rs
@@ -3587,7 +3587,6 @@ impl<'a> Checker<'a> {
     {
         let binding_index = self.bindings.len();
 
-        let mut overridden = None;
         if let Some((stack_index, scope_index)) = self
             .scope_stack
             .iter()
@@ -3619,7 +3618,6 @@ impl<'a> Checker<'a> {
                         | BindingKind::FutureImportation
                 );
                 if matches!(binding.kind, BindingKind::LoopVar) && existing_is_import {
-                    overridden = Some((*scope_index, *existing_binding_index));
                     if self.settings.rules.enabled(&Rule::ImportShadowedByLoopVar) {
                         self.diagnostics.push(Diagnostic::new(
                             violations::ImportShadowedByLoopVar(
@@ -3639,7 +3637,6 @@ impl<'a> Checker<'a> {
                                 cast::decorator_list(existing.source.as_ref().unwrap()),
                             ))
                     {
-                        overridden = Some((*scope_index, *existing_binding_index));
                         if self.settings.rules.enabled(&Rule::RedefinedWhileUnused) {
                             self.diagnostics.push(Diagnostic::new(
                                 violations::RedefinedWhileUnused(
@@ -3657,13 +3654,6 @@ impl<'a> Checker<'a> {
                         .push(binding_index);
                 }
             }
-        }
-
-        // If we're about to lose the binding, store it as overridden.
-        if let Some((scope_index, binding_index)) = overridden {
-            self.scopes[scope_index]
-                .overridden
-                .push((name, binding_index));
         }
 
         // Assume the rebound name is used as a global or within a loop.

--- a/src/rules/pyflakes/mod.rs
+++ b/src/rules/pyflakes/mod.rs
@@ -1170,19 +1170,11 @@ mod tests {
     fn aliased_import() -> Result<()> {
         flakes(
             "import fu as FU, bar as FU",
-            &[
-                Rule::UnusedImport,
-                Rule::RedefinedWhileUnused,
-                Rule::UnusedImport,
-            ],
+            &[Rule::RedefinedWhileUnused, Rule::UnusedImport],
         )?;
         flakes(
             "from moo import fu as FU, bar as FU",
-            &[
-                Rule::UnusedImport,
-                Rule::RedefinedWhileUnused,
-                Rule::UnusedImport,
-            ],
+            &[Rule::RedefinedWhileUnused, Rule::UnusedImport],
         )?;
 
         Ok(())
@@ -1219,18 +1211,9 @@ mod tests {
 
     #[test]
     fn redefined_while_unused() -> Result<()> {
-        flakes(
-            "import fu; fu = 3",
-            &[Rule::UnusedImport, Rule::RedefinedWhileUnused],
-        )?;
-        flakes(
-            "import fu; fu, bar = 3",
-            &[Rule::UnusedImport, Rule::RedefinedWhileUnused],
-        )?;
-        flakes(
-            "import fu; [fu, bar] = 3",
-            &[Rule::UnusedImport, Rule::RedefinedWhileUnused],
-        )?;
+        flakes("import fu; fu = 3", &[Rule::RedefinedWhileUnused])?;
+        flakes("import fu; fu, bar = 3", &[Rule::RedefinedWhileUnused])?;
+        flakes("import fu; [fu, bar] = 3", &[Rule::RedefinedWhileUnused])?;
 
         Ok(())
     }
@@ -1247,7 +1230,7 @@ mod tests {
             import os
         os.path
         "#,
-            &[Rule::UnusedImport, Rule::RedefinedWhileUnused],
+            &[Rule::RedefinedWhileUnused],
         )?;
 
         Ok(())
@@ -1285,7 +1268,7 @@ mod tests {
             pass
         os.path
         "#,
-            &[Rule::UnusedImport, Rule::RedefinedWhileUnused],
+            &[Rule::RedefinedWhileUnused],
         )?;
 
         Ok(())
@@ -1361,7 +1344,7 @@ mod tests {
             from bb import mixer
         mixer(123)
         "#,
-            &[Rule::UnusedImport, Rule::RedefinedWhileUnused],
+            &[Rule::RedefinedWhileUnused],
         )?;
 
         Ok(())
@@ -1433,7 +1416,7 @@ mod tests {
         def fu():
             pass
         "#,
-            &[Rule::UnusedImport, Rule::RedefinedWhileUnused],
+            &[Rule::RedefinedWhileUnused],
         )?;
 
         Ok(())
@@ -1511,7 +1494,7 @@ mod tests {
         class fu:
             pass
         "#,
-            &[Rule::UnusedImport, Rule::RedefinedWhileUnused],
+            &[Rule::RedefinedWhileUnused],
         )?;
 
         Ok(())
@@ -1775,7 +1758,7 @@ mod tests {
         for fu in range(2):
             pass
         "#,
-            &[Rule::UnusedImport, Rule::ImportShadowedByLoopVar],
+            &[Rule::ImportShadowedByLoopVar],
         )?;
 
         Ok(())
@@ -1932,11 +1915,7 @@ mod tests {
         try: pass
         except Exception as fu: pass
         "#,
-            &[
-                Rule::UnusedImport,
-                Rule::RedefinedWhileUnused,
-                Rule::UnusedVariable,
-            ],
+            &[Rule::RedefinedWhileUnused, Rule::UnusedVariable],
         )?;
 
         Ok(())
@@ -2246,7 +2225,7 @@ mod tests {
         import fu.bar, fu.bar
         fu.bar
         "#,
-            &[Rule::UnusedImport, Rule::RedefinedWhileUnused],
+            &[Rule::RedefinedWhileUnused],
         )?;
         flakes(
             r#"
@@ -2254,7 +2233,7 @@ mod tests {
         import fu.bar
         fu.bar
         "#,
-            &[Rule::UnusedImport, Rule::RedefinedWhileUnused],
+            &[Rule::RedefinedWhileUnused],
         )?;
 
         Ok(())
@@ -2410,7 +2389,7 @@ mod tests {
             fu
         fu
         "#,
-            &[Rule::UnusedImport, Rule::RedefinedWhileUnused],
+            &[Rule::RedefinedWhileUnused],
         )?;
 
         Ok(())

